### PR TITLE
[RFC] Remove some tv_list_item_…() functions

### PR DESCRIPTION
**Pull request by ZyX-I**
This is follow-up for #7708.

### CODE REVIEW

**Comment by ZyX-I** [comment]
Second time I got

```
[32m[2m[ RUN      ][0m[0m TermClose event triggers when fast-exiting terminal job stops: -- Output to stderr:
Vim: Caught deadly signal 'SIGHUP'

Vim: Finished.

CMake Error at /home/travis/build/neovim/neovim/cmake/RunTests.cmake:53 (message):
  Running functional tests failed with error: 1.
```

Is this a known failure?

File: src/nvim/lua/converter.c [hunk 1]
```diff
@@ -212,19 +212,27 @@ bool nlua_pop_typval(lua_State *lstate, typval_T *ret_tv)
...
+
             tv_list_append_list(cur.tv->vval.v_list, kv_pair);
-            listitem_T *const key = tv_list_item_alloc();
-            *TV_LIST_ITEM_TV(key) = decode_string(s, len, kTrue, false, false);
-            tv_list_append(kv_pair, key);
-            if (TV_LIST_ITEM_TV(key)->v_type == VAR_UNKNOWN) {
+            typval_T s_tv = decode_string(s, len, kTrue, false, false);
+            if (s_tv.v_type == VAR_UNKNOWN) {
               ret = false;
               tv_list_unref(kv_pair);
```
Comment (oni-link) [inline src/nvim/lua/converter.c]: Is it okay to free the list `kv_pair` but keep it as an item in the list `cur.tv->vval.v_list`?

File: src/nvim/eval.c [hunk 2]
```diff
@@ -1534,20 +1543,20 @@ ex_let_vars (
     EMSG(_("E688: More targets than List items"));
     return FAIL;
   }
-  // l may actually be NULL, but it should fail with E688 or even earlier if you
-  // try to do ":let [] = v:_null_list".
+  // lt may actually be NULL, but it should fail with E688 or even earlier if
```
Comment (oni-link) [inline src/nvim/eval.c]: Is `lt` a typo for `It` or should it stay `l`?

File: src/nvim/eval.c [hunk 3]
```diff
@@ -2887,28 +2898,25 @@ static int do_unlet_var(lval_T *const lp, char_u *const name_end, int forceit)
...
-      tv_list_item_remove(lp->ll_list, lp->ll_li);
       lp->ll_li = li;
       lp->ll_n1++;
+      if (lp->ll_li == NULL || (!lp->ll_empty2 && lp->ll_n2 < lp->ll_n1)) {
+        break;
+      } else {
+        last_li = lp->ll_li;
+      }
     }
+    tv_list_remove_items(lp->ll_list, first_li, last_li);
```
Comment (oni-link) [inline src/nvim/eval.c]: Can the range be empty?

File: src/nvim/eval.c [hunk 4]
```diff
@@ -12211,24 +12216,36 @@ static void find_some_match(typval_T *argvars, typval_T *rettv, int type)
...
-    tv_list_alloc_ret(rettv);
-    if (type == 4) {
+  switch (type) {
+    // matchlist(): return empty list when there are no matches.
+    case kSomeMatchList: {
+      tv_list_alloc_ret(rettv);
+      FALLTHROUGH;
+    }
+    // matchstrpos(): return ["", -1, -1, -1]
+    case kSomeMatchStrPos: {
```
Comment (oni-link) [inline src/nvim/eval.c]: In this case `tv_list_alloc_ret(rettv)` was not called.

File: src/nvim/eval.c [hunk 5]
```diff
@@ -12318,63 +12335,72 @@ static void find_some_match(typval_T *argvars, typval_T *rettv, int type)
...
+            if (type == kSomeMatchEnd) {
+              rettv->vval.v_number =
+                (varnumber_T)(regmatch.startp[0] - str);
+            } else {
+              rettv->vval.v_number =
+                (varnumber_T)(regmatch.endp[0] - str);
+            }
+            rettv->vval.v_number += (varnumber_T)(str - expr);
+          }
         }
```
Comment (oni-link) [inline src/nvim/eval.c]: Do we want to add a `break;` in case more cases are added?

File: src/nvim/eval.c [hunk 6]
```diff
@@ -12318,63 +12335,72 @@ static void find_some_match(typval_T *argvars, typval_T *rettv, int type)
...
-            (varnumber_T)(regmatch.startp[0] - str);
-        } else {
-          rettv->vval.v_number =
-            (varnumber_T)(regmatch.endp[0] - str);
+        case kSomeMatch:
+        case kSomeMatchEnd: {
+          if (l != NULL) {
+            rettv->vval.v_number = idx;
+          } else {
+            if (type == kSomeMatchEnd) {
```
Comment (oni-link) [inline src/nvim/eval.c]: Old value `0` is replaced with `kSomeMatchEnd` and `1` with `kSomeMatch` so 
`type != 0` should be transformed into `type !=kSomeMatchEnd`.

File: src/nvim/eval.c [hunk 7]
```diff
@@ -13131,16 +13155,13 @@ static void f_readfile(typval_T *argvars, typval_T *rettv, FunPtr fptr)
...
-      tv_list_item_remove(rettv->vval.v_list,
-                          tv_list_first(rettv->vval.v_list));
-      cnt--;
-    }
+  // For a negative line count use only the lines at the end of the file,
+  // free the rest.
+  if (maxline < -tv_list_len(rettv->vval.v_list)) {
+    listitem_T *const first_li = tv_list_find(rettv->vval.v_list, maxline);
+    listitem_T *const last_li = tv_list_last(rettv->vval.v_list);
+    tv_list_remove_items(rettv->vval.v_list, first_li, last_li);
```
Comment (oni-link) [inline src/nvim/eval.c]: Here the last items of the list are removed. In the old code the first items were removed.

**Comment by justinmk** [comment]
> Is this a known failure?

Yes it's a detected hang. Search for "hang" in the build log.

File: src/nvim/eval.c (hunk 2 above)
Comment (ZyX-I) [reply src/nvim/eval.c]: Oh, I thought that `l` there was a typo for `It`, but it appears I actually wrote variable name there.

File: src/nvim/eval.c (hunk 3 above)
Comment (ZyX-I) [reply src/nvim/eval.c]: This should produce E684.

**Comment by ZyX-I** [comment]
CI succeeded.

**Pull request merged**